.github/workflows/image-builder.yml

```diff
@@ -0,0 +1,210 @@
+      - name: Upload metadata
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: image-metadata-${{ inputs.directory}}-${{ matrix.image }}-${{ matrix.os }}
+          path: ${{ runner.temp }}/image-metadata/
+          if-no-files-found: error
+          retention-days: 1
+
+      # Upload the docker container so we can reference it in the next job.
+      - name: Upload tarball
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ inputs.directory }}-${{ matrix.image }}-${{ matrix.os }}-tarball
+          path: ${{ runner.temp }}/${{ matrix.image }}-${{ matrix.os }}.tar
+          if-no-files-found: error
+          retention-days: 1
+
+  build-images:
+    needs: [build-base-images]
+    if: |
+      always() && (needs.build-base-images.result == 'success' || needs.build-base-images.result == 'skipped')
+    strategy:
+      max-parallel: 5
+      matrix:
+        image: ${{ fromJSON(inputs.images) }}
+        os: [ubuntu-24.04, ubuntu-24.04-arm]
+        exclude: ${{ fromJSON(inputs.exclude) }}
+    runs-on: ${{ matrix.os }}
+    env:
+      CAN_PUSH_IMAGE: ${{ secrets.DOCKERHUB_USERNAME != '' }}
+
+    # These steps will be similar to `build-base-images` -- View the comments there for more details.
+    steps:
+      # Check out the repository with full depth so we can check if the workspace was modified
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check whether ${{ inputs.directory }}/${{ matrix.image }} was modified
+        id: check-modified
+        run: |
+          scripts/check_path_modified.sh ${{ inputs.directory }}/${{ matrix.image }} ${{ inputs.directory }}_${{ matrix.image }}_modified
+
+      - name: Login to DockerHub
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' && env.CAN_PUSH_IMAGE == 'true' }}
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Download tarball
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: actions/download-artifact@v4
+        # Download all base images that were built in the previous job
+        with:
+          pattern: '${{ inputs.directory }}-*-${{ matrix.os }}-tarball'
+          path: ${{ runner.temp }}
+          merge-multiple: true
+
+      - name: Tag base images
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        # Retag all base images that were downloaded in the previous step as latest.
+        # This is slightly simpler than using a named context.
+        run: |
+          for file in ${{ runner.temp }}/*-${{ matrix.os }}.tar; do
+            tagged_images=$(docker load -i "$file" | sed -n 's/Loaded image: //p')
```

Comment:
This `docker load` is the largest downside for us. For the prairielearn image to load `plbase`, that costs 20 seconds to download the tarball and 40 seconds to load the tarball. We only get this penalty on changes to `plbase` however.